op_crates/fetch/lib.rs

```diff
@@ -0,0 +1,262 @@
+use deno_core::url::Url;
+use deno_core::BufVec;
+use deno_core::JsRuntime;
+use deno_core::OpState;
+use deno_core::ZeroCopyBuf;
+use reqwest::header::HeaderMap;
+use reqwest::header::HeaderName;
+use reqwest::header::HeaderValue;
+use reqwest::header::USER_AGENT;
+use reqwest::redirect::Policy;
+use reqwest::Client;
+use reqwest::Method;
+use reqwest::Response;
+use serde::Deserialize;
+use std::cell::RefCell;
+use std::convert::From;
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+use std::path::PathBuf;
+use std::rc::Rc;
+
+pub fn init(isolate: &mut JsRuntime) {
+  let manifest_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
+  let files = vec![
+    manifest_dir.join("01_fetch_util.js"),
+    manifest_dir.join("03_dom_iterable.js"),
+    manifest_dir.join("11_streams.js"),
+    manifest_dir.join("20_headers.js"),
+    manifest_dir.join("26_fetch.js"),
+  ];
+  // TODO(nayeemrmn): https://github.com/rust-lang/cargo/issues/3946 to get the
+  // workspace root.
+  let display_root = manifest_dir.parent().unwrap().parent().unwrap();
+  for file in files {
+    println!("cargo:rerun-if-changed={}", file.display());
+    let display_path = file.strip_prefix(display_root).unwrap();
+    let display_path_str = display_path.display().to_string();
+    js_check(isolate.execute(
+      &("deno:".to_string() + &display_path_str.replace('\\', "/")),
+      &std::fs::read_to_string(&file).unwrap(),
+    ));
+  }
+}
+
+pub trait FetchPermissions {
+  fn check_net_url(&self, url: &Url) -> Result<(), AnyError>;
+  fn check_read(&self, p: &PathBuf) -> Result<(), AnyError>;
+}
+
+pub fn get_declaration() -> PathBuf {
+  PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("lib.deno_fetch.d.ts")
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+struct FetchArgs {
+  method: Option<String>,
+  url: String,
+  headers: Vec<(String, String)>,
+  client_rid: Option<u32>,
+}
+
+pub async fn op_fetch<FP>(
+  state: Rc<RefCell<OpState>>,
+  args: Value,
+  data: BufVec,
+) -> Result<Value, AnyError>
+where
+  FP: FetchPermissions + 'static,
+{
+  let args: FetchArgs = serde_json::from_value(args)?;
+  let url = args.url;
+
+  let client = if let Some(rid) = args.client_rid {
+    let state_ = state.borrow();
+    let r = state_
+      .resource_table
+      .get::<HttpClientResource>(rid)
+      .ok_or_else(bad_resource_id)?;
+    r.client.clone()
+  } else {
+    let state_ = state.borrow();
+    let client = state_.borrow::<reqwest::Client>();
+    client.clone()
+  };
+
+  let method = match args.method {
+    Some(method_str) => Method::from_bytes(method_str.as_bytes())?,
+    None => Method::GET,
+  };
+
+  let url_ = url::Url::parse(&url)?;
+
+  // Check scheme before asking for net permission
+  let scheme = url_.scheme();
+  if scheme != "http" && scheme != "https" {
+    return Err(type_error(format!("scheme '{}' not supported", scheme)));
+  }
+
+  {
+    let state_ = state.borrow();
+    let permissions = state_.borrow::<Rc<FP>>();
```

Comment:
had to put `Rc<FP>` here which is bad - it implcitily expects implementor to put `FP` in an Rc... However I couldn't figure out how to code that in the type